src/mpi/init/mpir_finalize.c

```diff
@@ -72,6 +72,10 @@ int MPIR_Finalize_impl(void)
     int rank = MPIR_Process.comm_world->rank;
 
     MPIR_INIT_LOCK;
+    MPIR_Process.init_counter--;
+    if (MPIR_Process.init_counter > 0) {
+        goto fn_exit;
+    }
 
```

Comment:
> If the user program makes mistake that num of MPI_Session_finalize calls > num of MPI_Session_init, ...

That is an error usage and we should detect and throw an error. Otherwise, the counter goes negative and it will mess up later `MPI_Session_init` -- this PR  can't deal with it but the standards allows it.

Your point is correct, and I'll add a error check for it. In fact, we probably should have a per-session counter to ensure the `MPI_Session_finalize` always matches the `MPI_Session_init`.

Reply:
Added code in `MPIR_Session_finalize_impl` to free the session object. Because we have validation code validate the session handle, `MPI_Session_finaize` will be matched to `MPI_Session_init`. User wouldn't be able to `MPI_Session_finalize` the same session twice because the second time the session object will be invalid.